core/meta/src/TClass.cxx

```diff
@@ -3455,38 +3455,32 @@ const char *TClass::GetSharedLibs()
 
 TList *TClass::GetListOfBases()
 {
-   if (!fBase) {
-      if (fCanLoadClassInfo) {
-         if (fState == kHasTClassInit) {
+   if (fBase)
+      return fBase;
 
-            R__LOCKGUARD(gInterpreterMutex);
-            // NOTE: Add test to prevent redo if another thread has already done the work.
-            // if (!fHasRootPcmInfo) {
+   if (fState == kHasTClassInit) {
+      R__LOCKGUARD(gInterpreterMutex);
 
-            // The bases are in our ProtoClass; we don't need the class info.
-            TProtoClass *proto = TClassTable::GetProtoNorm(GetName());
-            if (proto && proto->FillTClass(this)) {
-               // Not sure this code is still needed
-               // R__ASSERT(kFALSE);
+      // The bases are in our ProtoClass; we don't need the class info.
+      TProtoClass *proto = TClassTable::GetProtoNorm(GetName());
+      if (proto && proto->FillTClass(this))
+         fHasRootPcmInfo = true;
+   }
 
-               fHasRootPcmInfo = kTRUE;
-            }
-         }
-         // We test again on fCanLoadClassInfo has another thread may have executed it.
-         if (!fHasRootPcmInfo && !fCanLoadClassInfo) {
-            LoadClassInfo();
-         }
-      }
-      if (!fClassInfo) return 0;
+   if (fCanLoadClassInfo)
+      LoadClassInfo();
+
+   if (!fClassInfo)
+      return nullptr;
 
-      if (!gInterpreter)
-         Fatal("GetListOfBases", "gInterpreter not initialized");
+   if (!gInterpreter)
+      Fatal("GetListOfBases", "gInterpreter not initialized");
 
+   if (!fBase) {
       R__LOCKGUARD(gInterpreterMutex);
-      if(!fBase) {
-         gInterpreter->CreateListOfBaseClasses(this);
-      }
+      gInterpreter->CreateListOfBaseClasses(this);
```

Comment:
Actually TCling::CreateListOfBaseClasses takes the lock so it is not necessary to take it here.

If we were to take it then one would have to add a comment why the 'usual' pattern
if (needToBeSet) {
    TakeLock();
     if (needToBeSet) { // Redundant test is necessary since another thread may have set it while we were waiting
```
is not needed (and the answer is here because CreateListOfBaseClasses does the check.

Reply:
If we have a function called `TCling::CreateListOfBaseClasses()`, one has to wonder why this function isn't simply
```cpp
TList *TClass::GetListOfBases()
￼{
   if (!fBase && gInterpreter)
      gInterpreter->CreateListOfBaseClasses(this);
   return fBase;
}
```
or similar.

Reply:
I suppose that GetListOfBases is currently:
   * if needed change internal state to make base class information available
   * do a 'new ListOfBase'
but indeed I see your point ...